netty/src/main/java/io/grpc/netty/NettyClientTransport.java

```diff
@@ -194,6 +203,33 @@ public ClientStream newStream(
     if (channel == null) {
       return new FailingClientStream(statusExplainingWhyTheChannelIsNull, tracers);
     }
+    if (negotiator instanceof ClientTlsProtocolNegotiator && callOptions.getAuthority() != null) {
+      ClientTlsProtocolNegotiator clientTlsProtocolNegotiator =
+          (ClientTlsProtocolNegotiator) negotiator;
+      if (!clientTlsProtocolNegotiator.canVerifyAuthorityOverride()) {
+        return new FailingClientStream(Status.INTERNAL.withDescription(
+            "Can't allow authority override in rpc when X509ExtendedTrustManager is not available"),
+            tracers);
+      }
+      boolean peerVerified;
+      if (authoritiesAllowedForPeer.containsKey(callOptions.getAuthority())) {
+        peerVerified = authoritiesAllowedForPeer.get(callOptions.getAuthority());
+      } else {
+        try {
+          clientTlsProtocolNegotiator.verifyAuthorityAllowedForPeerCert(callOptions.getAuthority());
+          peerVerified = true;
+        } catch (SSLPeerUnverifiedException | CertificateException e) {
+          peerVerified = false;
+          logger.log(Level.FINE, "Peer hostname verification failed for authority '{}'.",
+              callOptions.getAuthority());
+        }
+        authoritiesAllowedForPeer.put(callOptions.getAuthority(), peerVerified);
+      }
+      if (!peerVerified) {
+        return new FailingClientStream(Status.INTERNAL.withDescription(
+            "Peer hostname verification failed for authority"), tracers);
```

Comment:
At the very least, we will want to include the failed authority in this error message. It'd probably be good to include `e` from that exception above as well.

Reply:
Done.